Join partition table with normal table will cause error: invalid memory address or nil pointer dereference

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table tp (id int primary key) partition by range (id) (partition p0 values less than (100));
create table tn (id int primary key);
insert into tp values(1),(2);
insert into tn values(1),(2);

select * from tp,tn where tp.id=tn.id and tn.id=1 for update;
```

### 2. What did you expect to see? (Required)

No error

### 3. What did you see instead (Required)

```
mysql> select * from tp,tn where tp.id=tn.id and tn.id=1 for update;
ERROR 1105 (HY000): runtime error: invalid memory address or nil pointer dereference
```

### 4. What is your TiDB version? (Required)

master

``` go
[2021/07/15 10:42:05.376 +08:00] [ERROR] [misc.go:89] ["panic in the recoverable goroutine"] [r="\"invalid memory address or nil pointer dereference\""] ["stack trace"="github.com/pingcap/tidb/util.WithRecovery.func1\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/util/misc.go:91\nruntime.gopanic\n\t/Users/xuhuaiyu/.gvm/gos/go1.16/src/runtime/panic.go:965\nruntime.panicmem\n\t/Users/xuhuaiyu/.gvm/gos/go1.16/src/runtime/panic.go:212\nruntime.sigpanic\n\t/Users/xuhuaiyu/.gvm/gos/go1.16/src/runtime/signal_unix.go:734\ngithub.com/pingcap/parser/model.(*ColumnInfo).GetOriginDefaultValue\n\t/Users/xuhuaiyu/Development/GOPATH/pkg/mod/github.com/pingcap/parser@v0.0.0-20210707071004-31c87e37af5c/model/model.go:158\ngithub.com/pingcap/tidb/table.GetColOriginDefaultValue\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/table/column.go:484\ngithub.com/pingcap/tidb/executor.NewRowDecoder.func2\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/builder.go:4167\ngithub.com/pingcap/tidb/util/rowcodec.(*ChunkDecoder).DecodeToChunk\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/util/rowcodec/decoder.go:240\ngithub.com/pingcap/tidb/executor.DecodeRowValToChunk\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/point_get.go:517\ngithub.com/pingcap/tidb/executor.(*PointGetExecutor).Next\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/point_get.go:301\ngithub.com/pingcap/tidb/executor.Next\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/executor.go:286\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).fetchBuildSideRows\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/join.go:268\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func2\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/join.go:702\ngithub.com/pingcap/tidb/util.WithRecovery\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/util/misc.go:94"]
[2021/07/15 10:42:05.378 +08:00] [INFO] [conn.go:995] ["command dispatched failed"] [conn=3] [connInfo="id:3, addr:127.0.0.1:54951 status:10, collation:utf8_general_ci, user:root"] [command=Query] [status="inTxn:0, autocommit:1"] [sql="select * from tp,tn where tp.id=tn.id and tn.id=1 for update"] [txn_mode=OPTIMISTIC] [err="runtime error: invalid memory address or nil pointer dereference\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func3\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/join.go:706\ngithub.com/pingcap/tidb/util.WithRecovery.func1\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/util/misc.go:86\nruntime.gopanic\n\t/Users/xuhuaiyu/.gvm/gos/go1.16/src/runtime/panic.go:965\nruntime.panicmem\n\t/Users/xuhuaiyu/.gvm/gos/go1.16/src/runtime/panic.go:212\nruntime.sigpanic\n\t/Users/xuhuaiyu/.gvm/gos/go1.16/src/runtime/signal_unix.go:734\ngithub.com/pingcap/parser/model.(*ColumnInfo).GetOriginDefaultValue\n\t/Users/xuhuaiyu/Development/GOPATH/pkg/mod/github.com/pingcap/parser@v0.0.0-20210707071004-31c87e37af5c/model/model.go:158\ngithub.com/pingcap/tidb/table.GetColOriginDefaultValue\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/table/column.go:484\ngithub.com/pingcap/tidb/executor.NewRowDecoder.func2\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/builder.go:4167\ngithub.com/pingcap/tidb/util/rowcodec.(*ChunkDecoder).DecodeToChunk\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/util/rowcodec/decoder.go:240\ngithub.com/pingcap/tidb/executor.DecodeRowValToChunk\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/point_get.go:517\ngithub.com/pingcap/tidb/executor.(*PointGetExecutor).Next\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/point_get.go:301\ngithub.com/pingcap/tidb/executor.Next\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/executor.go:286\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).fetchBuildSideRows\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/join.go:268\ngithub.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func2\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/join.go:702\ngithub.com/pingcap/tidb/util.WithRecovery\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/util/misc.go:94\nruntime.goexit\n\t/Users/xuhuaiyu/.gvm/gos/go1.16/src/runtime/asm_amd64.s:1371"]

```

--- a/planner/core/integration_test.go
+++ b/planner/core/integration_test.go
@@ -3970,3 +3970,13 @@ func (s *testIntegrationSerialSuite) TestSelectIgnoreTemporaryTableInView(c *C) 
 	tk.MustQuery("select * from v5").Check(testkit.Rows("1 2", "3 4"))
 
 }
+
+func (s *testIntegrationSerialSuite) TestIssue26250(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("create table tp (id int primary key) partition by range (id) (partition p0 values less than (100));")
+	tk.MustExec("create table tn (id int primary key);")
+	tk.MustExec("insert into tp values(1),(2);")
+	tk.MustExec("insert into tn values(1),(2);")
+	tk.MustQuery("select * from tp,tn where tp.id=tn.id and tn.id=1 for update;").Check(testkit.Rows("1 1"))
+}

--- a/planner/core/planbuilder.go
+++ b/planner/core/planbuilder.go
@@ -1234,6 +1234,10 @@ func (b *PlanBuilder) buildSelectLock(src LogicalPlan, lock *ast.SelectLockInfo)
 func addExtraPIDColumnToDataSource(p LogicalPlan, info *extraPIDInfo) error {
 	switch raw := p.(type) {
 	case *DataSource:
+		// Fix issue 26250, do not add extra pid column to normal table.
+		if raw.tableInfo.GetPartitionInfo() == nil {
+			return nil
+		}
 		raw.addExtraPIDColumn(info)
 		return nil
 	default:

